rmc/src/service.rs

```diff
@@ -0,0 +1,425 @@
+//! Reliable MultiCast - a primitive for Reliable Broadcast protocol.
+use crate::{
+    handler::Handler, scheduler::TaskScheduler, RmcHash, RmcIncomingMessage, RmcOutgoingMessage,
+    Task,
+};
+pub use aleph_bft_crypto::{
+    Indexed, MultiKeychain, Multisigned, NodeCount, PartialMultisignature, PartiallyMultisigned,
+    Signable, Signature, Signed, UncheckedSigned,
+};
+use aleph_bft_types::Terminator;
+use core::fmt::Debug;
+use futures::{
+    channel::mpsc::{UnboundedReceiver, UnboundedSender},
+    FutureExt, StreamExt,
+};
+use log::{debug, error, warn};
+use std::hash::Hash;
+
+/// Reliable Multicast Box
+///
+/// The instance of [`Service<H, MK>`] reliably broadcasts hashes of type `H`,
+/// and when a hash is successfully broadcast, the multisigned hash `Multisigned<H, MK>`
+/// is sent to the network.
+///
+/// A node with an instance of [`Service<H, MK>`] can initiate broadcasting a message `msg: H`
+/// by sending the [`RmcIncomingMessage::StartRmc`] variant through the network to [`Service<H, MK>`].
+/// As a result, the node signs `msg` and starts broadcasting the signed message via the network.
+/// When sufficintly many nodes initiate rmc with the same message `msg` and a node collects enough
+/// signatures to form a complete multisignature under the message, the multisigned message
+/// is sent back through the network by the instance of [`Service<H, MK>`].
+///
+/// We refer to the documentation https://cardinal-cryptography.github.io/AlephBFT/reliable_broadcast.html
+/// for a high-level description of this protocol and how it is used for fork alerts.
+pub struct Service<H: Signable + Hash, MK: MultiKeychain> {
+    network_rx: UnboundedReceiver<RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>>,
+    network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
+    scheduler: Box<dyn TaskScheduler<Task<H, MK>>>,
+}
+
+impl<H: Signable + Hash + Eq + Clone + Debug, MK: MultiKeychain> Service<H, MK> {
+    pub fn new(
+        network_rx: UnboundedReceiver<
+            RmcIncomingMessage<H, MK::Signature, MK::PartialMultisignature>,
+        >,
+        network_tx: UnboundedSender<RmcOutgoingMessage<H, MK>>,
```

Comment:
hmm, `Network` from `types` won't fit here as it assumes the sent data type is the same as received data type.

Reply:
Yeah, I'm suggesting it should return `RMCInterface` (with better name plx) rather than use any traits – we don't really use it as a network anyway, especially with the `start_rmc` thing.

Reply:
I'd need to create a separate interface for that. Also `IO` in `alerter` will still be needed.

Reply:
Yeah, something like that. The `IO` thing would've been a nice bonus, but it's not crucial.

Reply:
Tho, using a trait here seems too generic to me. I think imma just wrap these channels with a `struct RMCIO` or sth like that. This would be implemented with channels but it will obscure them from the user.

Reply:
Yeah, that's exactly what I had in mind, sorry for not being clear. ^^"

Reply:
I encountered some problems with this approach (dm-ed the details). Considering leaving the channels as they are...